eventstore/store_consolidated_billable_events_test.go

```diff
@@ -0,0 +1,359 @@
+
+		err = db.Schema.Consolidate(eventio.EventFilter{
+			RangeStart: "2001-01-01",
+			RangeStop:  "2001-02-01",
+		})
+
+		for month := 1; month <= 4; month++ {
+			filter := eventio.EventFilter{
+				RangeStart: fmt.Sprintf("2017-%02d-01", month),
+				RangeStop:  fmt.Sprintf("2017-%02d-01", month+1),
+			}
+			err = db.Schema.Consolidate(filter)
+			Expect(err).NotTo(HaveOccurred())
+
+			allBillableEvents, err := db.Schema.GetBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			allConsolidatedBillableEvents, err := db.Schema.GetConsolidatedBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(allBillableEvents).To(Equal(allConsolidatedBillableEvents))
+
+			filter.OrgGUIDs = []string{scenario.GetOrgGUID("org1")}
+			allOrg1BillableEvents, err := db.Schema.GetBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			allOrg1ConsolidatedBillableEvents, err := db.Schema.GetConsolidatedBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(allOrg1BillableEvents).To(Equal(allOrg1ConsolidatedBillableEvents))
+
+			filter.OrgGUIDs = []string{scenario.GetOrgGUID("org2")}
+			allOrg2BillableEvents, err := db.Schema.GetBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			allOrg2ConsolidatedBillableEvents, err := db.Schema.GetConsolidatedBillableEvents(filter)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(allOrg2BillableEvents).To(Equal(allOrg2ConsolidatedBillableEvents))
+		}
+	})
+
+	It("Should fail to GetBillableEvents if query range is not one month", func() {
+		db, err := scenario.Open(cfg)
+		Expect(err).ToNot(HaveOccurred())
+		defer db.Close()
+
+		err = db.Schema.Consolidate(eventio.EventFilter{
+			RangeStart: "2001-01-01",
+			RangeStop:  "2001-02-01",
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+		_, err = db.Schema.GetConsolidatedBillableEvents(eventio.EventFilter{
+			RangeStart: "2001-02-01",
+			RangeStop:  "2001-02-02",
+		})
+		Expect(err).To(MatchError("consolidation only works with ranges starting and ending on month boundaries"))
+
+		_, err = db.Schema.GetConsolidatedBillableEvents(eventio.EventFilter{
+			RangeStart: "2001-01-15",
+			RangeStop:  "2001-02-15",
+		})
+		Expect(err).To(MatchError("consolidation only works with ranges starting and ending on month boundaries"))
+	})
+})
+
+var _ = Describe("Consolidate", func() {
+	var (
+		cfg      eventstore.Config
+		scenario *testenv.TestScenario
+	)
+
+	BeforeEach(func() {
+		cfg = testenv.BasicConfig
+		scenario = testenv.NewTestScenario("2001-01-01T00:00")
+	})
+
+	It("Should fail to Consolidate if organisation filter provided", func() {
+		db, err := scenario.Open(cfg)
+		Expect(err).ToNot(HaveOccurred())
+		defer db.Close()
+
+		err = db.Schema.Consolidate(eventio.EventFilter{
+			RangeStart: "2001-01-02",
+			RangeStop:  "2001-02-02",
+			OrgGUIDs:   []string{"banana", "pear"},
+		})
+		Expect(err).To(MatchError(
+			"consolidate must be called without an organisations filter (i.e. for all orgs)",
```

Comment:
Potential future refactor: use a different struct for consolidation filters that doesn't include the possibility of this filter.

Reply:
:+1: TBH it is a quite adhoc thing. 

There is actually no reason to not allow do the consolidation with the OrgGUIDs, but if we do so, the consolidation key should include the list of OrgGUIDs for which it is valid, and that would make it more complicated.

To add a new type, would work but we would have to do some more work to convert types. We could use a compose type or a interface for the intervals to avoid this, but that would silently accept any filter even with orgGUIDs.  

Throwing an error was our way to be explicit saying that it is done for all orgs, and the guids will be ignored.

Reply:
Yeah. That's fine. It was just a random thought. Obviously going to be horrid to implement at this point in time and almost certainly not worth the effort.